spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestFilterPushDown.java

```diff
@@ -0,0 +1,465 @@
+    sql("REFRESH TABLE %s", tableName);
+    sql("INSERT INTO %s VALUES (2, 200, 'd2', 'sd2')", tableName);
+
+    // the filter can be pushed completely because all specs include identity(dep)
+    checkCompleteFilterPushdown(
+        "dep = 'd1'" /* query predicate */,
+        "dep IS NOT NULL, dep = 'd1'" /* Iceberg scan filters */,
+        ImmutableList.of(row(1, 100, "d1", "sd1")));
+
+    table.updateSpec().removeField("sub_dep").removeField("dep").commit();
+    sql("REFRESH TABLE %s", tableName);
+    sql("INSERT INTO %s VALUES (3, 300, 'd3', 'sd3')", tableName);
+
+    // the filter can't be pushed completely because not all specs include identity(dep)
+    checkFilterPushdown(
+        "dep = 'd1'" /* query predicate */,
+        "isnotnull(dep) AND (dep = d1)" /* post scan filter */,
+        "dep IS NOT NULL, dep = 'd1'" /* Iceberg scan filters */,
+        ImmutableList.of(row(1, 100, "d1", "sd1")));
+  }
+
+  @Test
+  public void testFilterPushdownWithSpecialFloatingPointPartitionValues() {
+    sql(
+        "CREATE TABLE %s (id INT, salary DOUBLE)" + "USING iceberg " + "PARTITIONED BY (salary)",
+        tableName);
+
+    sql("INSERT INTO %s VALUES (1, 100.5)", tableName);
+    sql("INSERT INTO %s VALUES (2, double('NaN'))", tableName);
+    sql("INSERT INTO %s VALUES (3, double('infinity'))", tableName);
+    sql("INSERT INTO %s VALUES (4, double('-infinity'))", tableName);
+
+    checkCompleteFilterPushdown(
+        "salary = 100.5" /* query predicate */,
+        "salary IS NOT NULL, salary = 100.5" /* Iceberg scan filters */,
+        ImmutableList.of(row(1, 100.5)));
+
+    checkCompleteFilterPushdown(
+        "salary = double('NaN')" /* query predicate */,
+        "salary IS NOT NULL, is_nan(salary)" /* Iceberg scan filters */,
+        ImmutableList.of(row(2, Double.NaN)));
+
+    checkCompleteFilterPushdown(
+        "salary != double('NaN')" /* query predicate */,
+        "salary IS NOT NULL, NOT (is_nan(salary))" /* Iceberg scan filters */,
+        ImmutableList.of(
+            row(1, 100.5), row(3, Double.POSITIVE_INFINITY), row(4, Double.NEGATIVE_INFINITY)));
+
+    checkCompleteFilterPushdown(
+        "salary = double('infinity')" /* query predicate */,
+        "salary IS NOT NULL, salary = Infinity" /* Iceberg scan filters */,
+        ImmutableList.of(row(3, Double.POSITIVE_INFINITY)));
+
+    checkCompleteFilterPushdown(
+        "salary = double('-infinity')" /* query predicate */,
+        "salary IS NOT NULL, salary = -Infinity" /* Iceberg scan filters */,
+        ImmutableList.of(row(4, Double.NEGATIVE_INFINITY)));
+  }
+
+  private void checkCompleteFilterPushdown(
+      String predicate, String pushedFilters, List<Object[]> expectedRows) {
+
+    checkFilterPushdown(predicate, null, pushedFilters, expectedRows);
+  }
+
+  private void checkFilterPushdown(
+      String predicate, String postScanFilter, String pushedFilters, List<Object[]> expectedRows) {
```

Comment:
this is where postScanFilter and pushedFilters as SparkFilters and IcebergFilters may help clarity

Reply:
I agree here as we exactly verify Spark and Iceberg filter representations. Changed.